<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Importar-Librerías" data-toc-modified-id="Importar-Librerías-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Importar Librerías</a></span></li><li><span><a href="#Funciones-y-Clases" data-toc-modified-id="Funciones-y-Clases-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Funciones y Clases</a></span></li><li><span><a href="#Preprocesamiento-de-Datos" data-toc-modified-id="Preprocesamiento-de-Datos-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Preprocesamiento de Datos</a></span><ul class="toc-item"><li><span><a href="#Construcción-del-Dataset-de-Productos" data-toc-modified-id="Construcción-del-Dataset-de-Productos-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Construcción del Dataset de Productos</a></span><ul class="toc-item"><li><span><a href="#Lectura-de-los-Datos" data-toc-modified-id="Lectura-de-los-Datos-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Lectura de los Datos</a></span></li><li><span><a href="#Transformación-de-los-Datos" data-toc-modified-id="Transformación-de-los-Datos-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Transformación de los Datos</a></span></li></ul></li><li><span><a href="#Selección-de-Predictores" data-toc-modified-id="Selección-de-Predictores-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Selección de Predictores</a></span></li><li><span><a href="#Estandarización,-OneHotEncoding-y-Train-Test-Split" data-toc-modified-id="Estandarización,-OneHotEncoding-y-Train-Test-Split-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Estandarización, OneHotEncoding y Train-Test Split</a></span></li></ul></li><li><span><a href="#Modelos-de-Predicción---Sold-Quantity" data-toc-modified-id="Modelos-de-Predicción---Sold-Quantity-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Modelos de Predicción - Sold Quantity</a></span><ul class="toc-item"><li><span><a href="#XGBoost-(Gradient-Boosting-Trees)" data-toc-modified-id="XGBoost-(Gradient-Boosting-Trees)-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>XGBoost (Gradient Boosting Trees)</a></span><ul class="toc-item"><li><span><a href="#Análisis-de-Resultados-y-Rendimiento" data-toc-modified-id="Análisis-de-Resultados-y-Rendimiento-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Análisis de Resultados y Rendimiento</a></span></li></ul></li><li><span><a href="#NNs-(Redes-Neuronales)" data-toc-modified-id="NNs-(Redes-Neuronales)-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>NNs (Redes Neuronales)</a></span><ul class="toc-item"><li><span><a href="#Análisis-de-Resultados-y-Rendimiento" data-toc-modified-id="Análisis-de-Resultados-y-Rendimiento-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Análisis de Resultados y Rendimiento</a></span></li></ul></li><li><span><a href="#Random-Forest-(Bosques-Aleatorios)" data-toc-modified-id="Random-Forest-(Bosques-Aleatorios)-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Random Forest (Bosques Aleatorios)</a></span><ul class="toc-item"><li><span><a href="#Análisis-de-Resultados-y-Rendimiento" data-toc-modified-id="Análisis-de-Resultados-y-Rendimiento-4.3.1"><span class="toc-item-num">4.3.1&nbsp;&nbsp;</span>Análisis de Resultados y Rendimiento</a></span></li></ul></li><li><span><a href="#Stacked-Model" data-toc-modified-id="Stacked-Model-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Stacked Model</a></span><ul class="toc-item"><li><span><a href="#Análisis-de-Resultados-y-Rendimiento" data-toc-modified-id="Análisis-de-Resultados-y-Rendimiento-4.4.1"><span class="toc-item-num">4.4.1&nbsp;&nbsp;</span>Análisis de Resultados y Rendimiento</a></span></li></ul></li></ul></li><li><span><a href="#Conclusiones-y-Reflexiones" data-toc-modified-id="Conclusiones-y-Reflexiones-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Conclusiones y Reflexiones</a></span></li></ul></div>

# Sold Quantity Prediction

El propósito de este notebook es construir una serie de modelos de predicción para estimar la variable `ITEM_SOLD_QUANTITY` de un item de [Mercado Libre](https://www.mercadolibre.com.co/) con el dataset construido en [`MeliMLChallenge/Notebook/ApiPullingData.ipynb`](https://github.com/juanse1608/MeliMLChallenge/blob/main/Notebooks/ApiPullingData.ipynb).

Se recuerda que el dataset consta de __14574__ items (tratanto de tener la misma cantidad para cada una de las categorías de ítems de Meli Colombia) y __50__ variables asociadas al ítem, su vendedor y su envío.

## Importar Librerías

In [1]:
# Cambia el ancho de las celdas
from IPython.display import HTML, display
display(HTML(data="""<style>div#notebook-container{width: 60%;}div#menubar-container{width: 65%;} div#maintoolbar-container{width: 99%;}</style>"""))

In [2]:
# Procesamiento de datos
import os
import re
import pandas as pd
import numpy as np
import datetime as dt
import itertools as it
import requests

# Visualización
import matplotlib.pyplot as plt
import matplotlib.font_manager
from matplotlib.ticker import PercentFormatter
import chart_studio.plotly as py
import plotly.graph_objs as go
import plotly.express as px
from plotly.subplots import make_subplots

# ML y DS
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV, KFold
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import StackingRegressor
from sklearn.utils.multiclass import unique_labels
from sklearn.metrics import confusion_matrix
from xgboost import XGBRegressor
from xgboost import XGBRFRegressor
import tensorflow as tf

## Funciones y Clases

Esta sección tiene las funciones y clases creadas que se usarán a los largo de este notebook.

In [3]:
# Funcion que en encuentra el valor mas cercano de un array a un valor 
def find_nearest(array, value):
    array = np.asarray(array)
    idx = (np.abs(array - value)).argmin()
    return array[idx]

In [4]:
#Let's plot the confusion matrix
#Function which objective is to plot in a fancy way the confusion matrix
def plot_confusion_matrix(y_true, y_pred, classes,
                          normalize=False,
                          title=None,
                          cmap=plt.cm.Blues):
    """
    This function prints and plots the confusion matrix.
    Normalization can be applied by setting `normalize=True`.
    """
    if not title:
        if normalize:
            title = 'Normalized confusion matrix'
        else:
            title = 'Confusion matrix, without normalization'

    # Compute confusion matrix
    cm = confusion_matrix(y_true, y_pred)
    # Only use the labels that appear in the data
#     classes = classes[unique_labels(y_true, y_pred)]
    if normalize:
        cm = cm.astype('float') / cm.sum(axis=1)[:, np.newaxis]
        #print("Normalized confusion matrix")

    #print(cm)

    fig, ax = plt.subplots(figsize = (10,10))
    im = ax.imshow(cm, interpolation='nearest', cmap=cmap)
    ax.figure.colorbar(im, ax=ax)
    # We want to show all ticks...
    ax.set(xticks=np.arange(cm.shape[1]),
           yticks=np.arange(cm.shape[0]),
           # ... and label them with the respective list entries
           xticklabels=classes, yticklabels=classes,
           title=title,
           ylabel='True label',
           xlabel='Predicted label')

    # Rotate the tick labels and set their alignment.
    plt.setp(ax.get_xticklabels(), rotation=0, ha="right",
             rotation_mode="anchor")

    # Loop over data dimensions and create text annotations.
    fmt = '.2f' if normalize else 'd'
    thresh = cm.max() / 2.
    for i in range(cm.shape[0]):
        for j in range(cm.shape[1]):
            ax.text(j, i, format(cm[i, j], fmt),
                    ha="center", va="center",
                    color="white" if cm[i, j] > thresh else "black")
    fig.tight_layout()
    return ax

## Preprocesamiento de Datos

En esta sección se realiza un proceso general de limpieza y organización de los datos previo a la exploración de los datos.

### Construcción del Dataset de Productos

#### Lectura de los Datos
 
En esta subsección se leen el dataset de productos, se imprimen sus dimensiones, sus columnas y se mencionan sus tipos de variables.

In [5]:
# Se lee el .csv con el datset construido
products = pd.read_csv('../Data/Datasets/PRODUCTOS.csv')
products.columns = products.columns.str.upper() # Pone los nombres de las columnas en mayusculas
print('LAS DIMENSIONES DE LA BASE DE PRODUCTOS SON: {}'.format(products.shape))

LAS DIMENSIONES DE LA BASE DE PRODUCTOS SON: (14574, 50)


In [6]:
# Printea la columnas del dataset
products.columns

Index(['ITEM_ID', 'ITEM_SITE_ID', 'ITEM_TITLE', 'ITEM_PRICE',
       'ITEM_SALE_PRICE', 'ITEM_CURRENCY_ID', 'ITEM_AVAILABLE_QUANTITY',
       'ITEM_SOLD_QUANTITY', 'ITEM_BUYING_MODE', 'ITEM_LISTING_TYPE_ID',
       'ITEM_STOP_TIME', 'ITEM_CONDITION', 'ITEM_PERMALINK', 'ITEM_THUMBNAIL',
       'ITEM_THUMBNAIL_ID', 'ITEM_ACCEPTS_MERCADOPAGO', 'ITEM_ORIGINAL_PRICE',
       'ITEM_CATEGORY_ID', 'ITEM_OFFICIAL_STORE_ID', 'ITEM_DOMAIN_ID',
       'ITEM_CATALOG_PRODUCT_ID', 'ITEM_ORDER_BACKEND',
       'ITEM_USE_THUMBNAIL_ID', 'SEARCH_CATEGORY_ID', 'SEARCH_CATEGORY_NAME',
       'SEARCH_OFFSET', 'SELLER_ID', 'SELLER_REP_TRANSACTIONS_TOTAL',
       'SELLER_REP_TRANSACTIONS_CANCELED', 'SELLER_REP_RATING_NEG',
       'SELLER_REP_RATING_POS', 'SELLER_REP_RATING_NEU',
       'SELLER_TRANSACTIONS_COMPLETED', 'SELLER_STATUS',
       'SELLER_METRICS_CLAIMS_RATE', 'SELLER_METRICS_CLAIMS_VALUE',
       'SELLER_METRICS_CLAIMS_PERIOD', 'SELLER_METRICS_DELAY_RATE',
       'SELLER_METRICS_DELAY_VALUE', 'SEL

Se puede ver claramente que se encuentran los siguientes tipos de variables:
    
1. Tipo `ITEM`: Variables asociadas al ítem.


2. Tipo `SELLER`: Variables asociadas al vendedor.


3. Tipo `SHIPPING`: Variables asociadas al envío.


4. Tipo `ADRESS`: Variable que indica el departamento donde se encuentra el ítem.


5. Tipo `SEARCH`: Variables asociadas a la búsqueda: categoría y offset.

#### Transformación de los Datos

En esta subsección se arreglan, limpian y transforman el dataset de productos.

In [7]:
# Arreglo a la ADRESS_STATE_ID: para algunos productos el ADRESS_STATE_ID hay que arreglar su dato
# Se uso el ITEM_PERMALINK para entender cual era el verdadero dato de ADRESS_STATE_ID
cambios_de_direccion = {
                        'TUNPUEJPR1gxMDljZA': 'CO-DC', 
                        'TUNPUEFOVGFiZWI3': 'CO-ANT ',
                        'TUNPUENVTmE3NmQ4': 'CO-CUN',
                        'TUNPUFZBTGExNmNjNg': 'CO-VAC',
                        'TUNPUFJJU2ExMWIyYg': 'CO-RIS',
                        'TUNPUEFUTG9mNDk5': 'CO-ATL',
                        'TUNPUFRPTGExNGZkNA': 'CO-TOL',
                        'TUNPUFNBTnJlMjMw': 'CO-SAN',
                        'TUNPUEJPTHI1Mzlk': 'CO-BOL',
                        'TUNPUE1BR2FiZjQ0': 'CO-MAG',
                        'TUNPUENBTHNjODY4': 'CO-CAL',
                        'TUNPUE1FVGExNzFjNQ': 'CO-MET',
                        'TUNPUFFVSW9kYmZm': 'CO-QUI',
                        'TUNPUE5PUnIxNDkyZg': 'CO-NSA', 
                        'TUNPUEFSQ2E4Zjc3': 'CO-SAP',
                        'TUNPUEJPWWE4YzMz': 'CO-BOY',
                        'TUNPUENBVWExM2Q1NQ': 'CO-CAU',
                        'TUNPUENPUmFkZGIw': 'CO-COR',
                        'TUNPUEdVQWExOTYx': 'CO-LAG',
                        'TUNPUE5BUm8xYzk4': 'CO-NAR', 
                        'TUNPUFNVQ2U4ZWQ0': 'CO-SUC',
                        'TUNPUENBU2U2OWIy': 'CO-CAS',
                        'TUNPUENFU3IxODA4Mg': 'CO-CES',
                        'TUNPUEFNQXMxMzQ2YQ': 'CO-AMA'
                       }

# Arregla un error que se detecto en las direcciones
products.loc[products['ADRESS_STATE_ID'].isin(cambios_de_direccion.keys()), 'ADRESS_STATE_ID'] = \
products.loc[products['ADRESS_STATE_ID'].isin(cambios_de_direccion.keys()), 'ADRESS_STATE_ID'].map(cambios_de_direccion)

In [8]:
## Ajuste a la condicion del item

# Se cambia nan por not_specified
products.loc[products['ITEM_CONDITION'].isna(), 'ITEM_CONDITION'] = 'not_specified'

# Ajustes al precio 

# Se quitan productos que no tienen precio (generalmente precio a convenir)
# Es decir este analisis aplica para productos con precio fijo
products = products.loc[~products['ITEM_PRICE'].isna()]
products.reset_index(drop=True, inplace=True)

# Se encontro en google la tasa de cambio de dolar a cop el dia 7 de marzo (cuando se descargaron los datos)
exchange_rate_usd_to_cop = 3142.99 
products.loc[products['ITEM_CURRENCY_ID'] == 'USD', 'ITEM_PRICE'] *= exchange_rate_usd_to_cop
products.loc[products['ITEM_CURRENCY_ID'] == 'USD', 'ITEM_CURRENCY_ID'] = 'COP'

# Cambia ITEM_ORIGINAL_PRICE por el ITEM_PRICE como se indica en el documento
products.loc[products['ITEM_ORIGINAL_PRICE'].isna(), 'ITEM_ORIGINAL_PRICE'] = \
products.loc[products['ITEM_ORIGINAL_PRICE'].isna(), 'ITEM_PRICE']

# Crea las variables de descuento absoluto y la tasa
products['ITEM_DISCOUNT_VALUE'] = products['ITEM_ORIGINAL_PRICE'] - products['ITEM_PRICE']
products['ITEM_DISCOUNT_RATE'] = products['ITEM_DISCOUNT_VALUE']/products['ITEM_ORIGINAL_PRICE']

# Crea la variable de venta total: This is an approximation
products['ITEM_SELL_TOTAL'] = products['ITEM_PRICE']*products['ITEM_SOLD_QUANTITY']

In [9]:
## Ajuste a variables del vendedor

# Cambio de nan to not_specified para SELLER_STATUS
products.loc[products['SELLER_STATUS'].isna(), 'SELLER_STATUS'] = 'not_specified'
products.loc[products['SELLER_LEVEL_ID'].isna(), 'SELLER_LEVEL_ID'] = 'not_specified'

# Cambio nan to 0 para SELLER_METRICS_SALES_PERIOD
products.loc[products['SELLER_METRICS_SALES_PERIOD'].isna(), 'SELLER_METRICS_SALES_COMPLETED'] = 0
products.loc[products['SELLER_METRICS_SALES_PERIOD'].isna(), 'SELLER_METRICS_SALES_PERIOD'] = '60 days'
products.loc[products['SELLER_METRICS_SALES_PERIOD'] == '60 months', 'SELLER_METRICS_SALES_COMPLETED'] *= (60/(30*60))
products.loc[products['SELLER_METRICS_SALES_PERIOD'] == '3 months', 'SELLER_METRICS_SALES_COMPLETED'] *= (60/90)
products.loc[products['SELLER_METRICS_SALES_PERIOD'] == '365 days', 'SELLER_METRICS_SALES_COMPLETED'] *= (60/365)
products['SELLER_METRICS_SALES_PERIOD'] = '60 days'

# Estandarizacion de las metricas absolutas del vendedor (por periodo de tiempo)

# Cancelaciones 
products.loc[products['SELLER_METRICS_CANCELLATIONS_PERIOD'] == '60 months', 'SELLER_METRICS_CANCELLATIONS_VALUE'] *= (60/(30*60))
products.loc[products['SELLER_METRICS_CANCELLATIONS_PERIOD'] == '3 months', 'SELLER_METRICS_CANCELLATIONS_VALUE'] *= (60/90)
products.loc[products['SELLER_METRICS_CANCELLATIONS_PERIOD'] == '365 days', 'SELLER_METRICS_CANCELLATIONS_VALUE'] *= (60/365)
products['SELLER_METRICS_CANCELLATIONS_PERIOD'] = '60 days'

# Quejas 
products.loc[products['SELLER_METRICS_CLAIMS_PERIOD'] == '60 months', 'SELLER_METRICS_CLAIMS_VALUE'] *= (60/(30*60))
products.loc[products['SELLER_METRICS_CLAIMS_PERIOD'] == '3 months', 'SELLER_METRICS_CLAIMS_VALUE'] *= (60/90)
products.loc[products['SELLER_METRICS_CLAIMS_PERIOD'] == '365 days', 'SELLER_METRICS_CLAIMS_VALUE'] *= (60/365)
products['SELLER_METRICS_CLAIMS_PERIOD'] = '60 days'

# Demoras 
products.loc[products['SELLER_METRICS_DELAY_PERIOD'] == '60 months', 'SELLER_METRICS_DELAY_VALUE'] *= (60/(30*60))
products.loc[products['SELLER_METRICS_DELAY_PERIOD'] == '3 months', 'SELLER_METRICS_DELAY_VALUE'] *= (60/90)
products.loc[products['SELLER_METRICS_DELAY_PERIOD'] == '365 days', 'SELLER_METRICS_DELAY_VALUE'] *= (60/365)
products['SELLER_METRICS_DELAY_PERIOD'] = '60 days'

## Ajustes a variables del shipping/envio

# Cambio de nan to not_specified para SHIPPING_LOGISTIC_TYPE
products.loc[products['SHIPPING_LOGISTIC_TYPE'].isna(), 'SHIPPING_LOGISTIC_TYPE'] = 'not_specified'

In [10]:
# Encuentra las categorias cuyas cantidades vendidas son 0 y las remueve
sold_quantity_per_category = products.groupby('SEARCH_CATEGORY_ID')['ITEM_SELL_TOTAL'].sum().sort_values(ascending=False)
categories_zero_sold_quantity = sold_quantity_per_category.loc[sold_quantity_per_category == 0].index.tolist()
products = products.loc[~products['SEARCH_CATEGORY_ID'].isin(categories_zero_sold_quantity)]

# Se remueven los productos cuya cantidad vendida es de 5000
products = products.loc[products['ITEM_SOLD_QUANTITY'] < 5000]
products = products.loc[~products['ITEM_DOMAIN_ID'].isna()]

# Imprime las dimensiones del dataset de products final
print('LAS DIMENSIONES DE LA BASE DE PRODUCTOS SON: {}'.format(products.shape))

LAS DIMENSIONES DE LA BASE DE PRODUCTOS SON: (13043, 53)


### Selección de Predictores

In [11]:
# Variables de item y search  
item_vars = [
    'ITEM_PRICE',
    'ITEM_AVAILABLE_QUANTITY',
    'ITEM_SOLD_QUANTITY', # Variable de respuesta
    'ITEM_LISTING_TYPE_ID',
    'ITEM_CONDITION',
    'ITEM_DOMAIN_ID',
    'ITEM_ACCEPTS_MERCADOPAGO',
    'ITEM_ORIGINAL_PRICE',
    'ITEM_USE_THUMBNAIL_ID',
    'SEARCH_CATEGORY_ID', # La categoria del busqueda del item 
    'SEARCH_OFFSET', # El offset de busqueda que en el que se obtuvo el item
#     'ITEM_DISCOUNT_VALUE', # Se omite ya que es una combinacion lineal de ITEM_PRICE y ITEM_ORIGINAL_PRICE
    'ITEM_DISCOUNT_RATE',
]

# Variables del seller y el shipping
seller_vars = [
    'SELLER_REP_TRANSACTIONS_TOTAL',
    'SELLER_REP_TRANSACTIONS_CANCELED',
    'SELLER_REP_RATING_NEG',
    'SELLER_REP_RATING_POS',
    'SELLER_REP_RATING_NEU',
    'SELLER_TRANSACTIONS_COMPLETED',
    'SELLER_METRICS_CLAIMS_RATE',
    'SELLER_METRICS_CLAIMS_VALUE',
    'SELLER_METRICS_DELAY_RATE',
    'SELLER_METRICS_DELAY_VALUE', 
    'SELLER_METRICS_SALES_COMPLETED',
    'SELLER_METRICS_CANCELLATIONS_RATE',
    'SELLER_METRICS_CANCELLATIONS_VALUE', 
    'SELLER_LEVEL_ID',
    'SHIPPING_FREE',
    'SHIPPING_LOGISTIC_TYPE', 
    'SHIPPING_PICK_UP',
    'ADRESS_STATE_ID', 
]

# Seleeciona unicamente columnas anteriormente definidas
products_selected = products[item_vars + seller_vars]
products_selected.reset_index(drop=True, inplace=True)
print('LAS DIMENSIONES DE LA DATA DE PRODUCTOS SON: {}'.format(products_selected.shape))

LAS DIMENSIONES DE LA DATA DE PRODUCTOS SON: (13043, 30)


### Estandarización, OneHotEncoding y Train-Test Split

In [12]:
# Crea un dataframe de variables numericas y otro de categoricas
numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
products_numeric = products_selected.select_dtypes(include=numerics)

categorical = ['object']
products_categorical = products_selected.select_dtypes(include=categorical)

# Cambia las variables 
binary = ['bool']
products_binary = products_selected.select_dtypes(include=binary)

# Revisa que todas las variables de products_selected esten en products_numeric, products_binary y products_categorical
assert(products_selected.shape[1] == (products_categorical.shape[1]+products_numeric.shape[1]+products_binary.shape[1]))

# Imprime las dimensiones de cada base
print('LAS DIMENSIONES DE LA DATA NUMERICA DE PRODUCTOS SON: {}'.format(products_numeric.shape))
print('LAS DIMENSIONES DE LA DATA CATEGORICA DE PRODUCTOS SON: {}'.format(products_categorical.shape))
print('LAS DIMENSIONES DE LA DATA BINARIA DE PRODUCTOS SON: {}'.format(products_binary.shape))

LAS DIMENSIONES DE LA DATA NUMERICA DE PRODUCTOS SON: (13043, 19)
LAS DIMENSIONES DE LA DATA CATEGORICA DE PRODUCTOS SON: (13043, 7)
LAS DIMENSIONES DE LA DATA BINARIA DE PRODUCTOS SON: (13043, 4)


In [13]:
## Cambio a 1-0 para las binarias
# Cambia las variables binarias True-False por 1-0 
products_binary = products_binary.astype(int)

## OneHotEncoding para las columnas categoricas
# Define el one_hot_encoder
one_hot_encoder = OneHotEncoder(sparse=False)
# Transforma los datos
products_encoded_categoricals = one_hot_encoder.fit_transform(products_categorical)

In [14]:
# Divide los datos en entrenamiento y prueba
Y = products_numeric['ITEM_SOLD_QUANTITY']
X = products_numeric.drop(columns='ITEM_SOLD_QUANTITY')
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.3, random_state=1608)
indexes_train = X_train.index.tolist()
indexes_test = X_test.index.tolist()

## Estandarizacion para las numericas
standard_scaler = StandardScaler() 
X_train_std = standard_scaler.fit_transform(X_train)
X_test_std = standard_scaler.transform(X_test)

## Construye la matriz de datos con todos los tipos de data
X_train_total = np.concatenate((X_train_std, products_encoded_categoricals[indexes_train,:],
                                products_binary.loc[indexes_train]), axis=1)
X_test_total = np.concatenate((X_test_std, products_encoded_categoricals[indexes_test,:],
                                products_binary.loc[indexes_test]), axis=1)

In [19]:
# Conteo de valores para la respuesta
Y_train.value_counts().reset_index().rename(columns={'index': 'ITEM_SOLD_QUANTITY',
                                                     'ITEM_SOLD_QUANTITY': 'COUNT'})

ITEM_SOLD_QUANTITY  COUNT
0                  500   1460
1                    5   1413
2                   50   1300
3                  250   1163
4                   25   1058
5                  100    778
6                  150    554
7                  200    426
8                    0    361
9                    1    211
10                   3    148
11                   4    122
12                   2    117
13                   9      3
14                  12      2
15                  14      2
16                  15      2
17                  59      1
18                   8      1
19                  35      1
20                  10      1
21                  13      1
22                  21      1
23                   6      1
24                  23      1
25                  31      1
26                  55      1

Es claro que algunos valores son muy escasos y hay otros que se encuentran mucho más presentes en los ítems tomados.

## Modelos de Predicción - Sold Quantity

En esta sección se construye y evalua el rendimiento de disintos modelos de Machine Learning para estimar la cantidad vendida de un ítem en función de las variables escogidas previamente.

### XGBoost (Gradient Boosting Trees)

Uno de los algoritmos que escogí para usar es Gradient Boosting Trees a través del framework `XGBoost` conocido por su excelente rendimiento y rapidez. Lo que se hara es buscar los mejores hiper-parámetros por validación cruzada y evaluar el desempeño del modelo.

In [14]:
# Parametros del XGBoost para tunear (se escogio el dart booster para evitar el overfitting)
param_tuning = {
        'learning_rate': [0.25],
        'max_depth': [5, 7],
        'min_child_weight': [3, 5],
        'subsample': [1.0],
        'colsample_bytree': [0.50],
        'n_estimators' : [100],
        'objective': ['reg:squarederror'],
        'booster': ['dart'],
        'rate_drop': [0.1, 0.25],
        'skip_drop': [0.1, 0.5]
    }

# Se evalua para todas las combinaciones de hiperparametros cual tiene el menor error de validacion
# usando validacion cruzada para 5-folds

# Modelo XGBRegressor
xgb_model = XGBRegressor()

# Validacion cruzada
k_folds = KFold(n_splits=5, shuffle=True, random_state=1608)
grid_search_cv = GridSearchCV(estimator = xgb_model,
                              param_grid = param_tuning,            
                              cv = k_folds,
                              n_jobs = -1,
                              verbose = 1)
cv_fit = grid_search_cv.fit(X=X_train_total, y=Y_train) # Fit

Fitting 5 folds for each of 16 candidates, totalling 80 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 16 concurrent workers.
[Parallel(n_jobs=-1)]: Done  18 tasks      | elapsed:  1.5min
[Parallel(n_jobs=-1)]: Done  80 out of  80 | elapsed:  4.7min finished


In [15]:
# Imprime los mejores parametros bajo el score de cv calculado para cada iteracion
grid_search_cv.best_params_

{'booster': 'dart',
 'colsample_bytree': 0.5,
 'learning_rate': 0.25,
 'max_depth': 7,
 'min_child_weight': 5,
 'n_estimators': 100,
 'objective': 'reg:squarederror',
 'rate_drop': 0.1,
 'skip_drop': 0.5,
 'subsample': 1.0}

In [16]:
# Se generan las predicciones para el entrenamiento y prueba y se evaluan dos metricas: RMSE y MAE
xgb_best_model = grid_search_cv.best_estimator_
test_predictions = xgb_best_model.predict(X_test_total)
train_predictions = xgb_best_model.predict(X_train_total)

In [17]:
# Comparacion RSME train y test
rmse_train = np.sqrt(mean_squared_error(y_pred=train_predictions, y_true=Y_train))
rmse_test = np.sqrt(mean_squared_error(y_pred=test_predictions, y_true=Y_test))
print('RMSE DE ENTRENAMIENTO: {}'.format(np.round(rmse_train,2)))
print('RMSE DE PRUEBA: {}'.format(np.round(rmse_test,2)))
diferencia_porcentual = np.round(100*(rmse_test-rmse_train)/rmse_train,2)
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

RMSE DE ENTRENAMIENTO: 79.35
RMSE DE PRUEBA: 104.87
DIFERENCIA DE: 32.17%


In [18]:
# Comparacion MAE train y test
mae_train = mean_absolute_error(y_pred=train_predictions, y_true=Y_train)
mae_test = mean_absolute_error(y_pred=test_predictions, y_true=Y_test)
diferencia_porcentual = np.round(100*(mae_test-mae_train)/mae_train,2)
print('MAE DE ENTRENAMIENTO: {}'.format(np.round(mae_train,2)))
print('MAE DE PRUEBA: {}'.format(np.round(mae_test,2)))
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

MAE DE ENTRENAMIENTO: 54.74
MAE DE PRUEBA: 72.84
DIFERENCIA DE: 33.07%


In [19]:
values = sorted(Y_train.unique())
test_predictions = [find_nearest(values, p)for p in test_predictions] 
train_predictions = [find_nearest(values, p)for p in train_predictions] 

In [20]:
# Comparacion RSME train y test
rmse_train = np.sqrt(mean_squared_error(y_pred=train_predictions, y_true=Y_train))
rmse_test = np.sqrt(mean_squared_error(y_pred=test_predictions, y_true=Y_test))
print('RMSE DE ENTRENAMIENTO: {}'.format(np.round(rmse_train,2)))
print('RMSE DE PRUEBA: {}'.format(np.round(rmse_test,2)))
diferencia_porcentual = np.round(100*(rmse_test-rmse_train)/rmse_train,2)
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

RMSE DE ENTRENAMIENTO: 86.3
RMSE DE PRUEBA: 111.63
DIFERENCIA DE: 29.35%


In [21]:
# Comparacion MAE train y test
mae_train = mean_absolute_error(y_pred=train_predictions, y_true=Y_train)
mae_test = mean_absolute_error(y_pred=test_predictions, y_true=Y_test)
diferencia_porcentual = np.round(100*(mae_test-mae_train)/mae_train,2)
print('MAE DE ENTRENAMIENTO: {}'.format(np.round(mae_train,2)))
print('MAE DE PRUEBA: {}'.format(np.round(mae_test,2)))
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

MAE DE ENTRENAMIENTO: 49.49
MAE DE PRUEBA: 69.39
DIFERENCIA DE: 40.21%


In [22]:
# Se crea un dataframe con los resultados
resultados_train = pd.DataFrame({'TYPE': ['TRAIN']*len(Y_train), 'TRUE': Y_train, 'PREDICTED': train_predictions})
resultados_test = pd.DataFrame({'TYPE': ['TEST']*len(Y_test), 'TRUE': Y_test, 'PREDICTED': test_predictions})
resultados_xgb = pd.concat([resultados_train, resultados_test], axis=0).reset_index(drop=True)
resultados_xgb['ABSOLUTE_ERROR'] = (resultados_xgb['TRUE']-resultados_xgb['PREDICTED']).abs()
resultados_xgb['SQUARE_ERROR'] = (resultados_xgb['TRUE']-resultados_xgb['PREDICTED'])**2

In [23]:
# Resultados por cantidad
resultados_por_cantidad = resultados_xgb.groupby(['TYPE', 'TRUE'])[['ABSOLUTE_ERROR', 'SQUARE_ERROR']].mean().reset_index()

In [24]:
# Imprime los resultados por cantidad en cuanto al MAE y MSE para los datos de prueba
resultados_por_cantidad.loc[resultados_por_cantidad['TYPE']=='TEST']

TYPE  TRUE  ABSOLUTE_ERROR  SQUARE_ERROR
0   TEST     0       13.389937    509.805031
1   TEST     1       14.776119    718.059701
2   TEST     2       11.974026    381.818182
3   TEST     3       18.562500    864.020833
4   TEST     4       15.784615    673.815385
5   TEST     5       38.708943   3173.508943
6   TEST     6       69.000000   5386.000000
7   TEST     7       52.000000   2704.000000
8   TEST     8        8.000000     64.000000
9   TEST    14       11.000000    121.000000
10  TEST    25       54.049107   5547.450893
11  TEST    50       63.566004   7161.996383
12  TEST    56       44.000000   1936.000000
13  TEST   100       66.703170   8505.406340
14  TEST   150       61.799127   6986.052402
15  TEST   200       62.897436   7059.512821
16  TEST   250       64.666667  11286.886406
17  TEST   500      165.544554  46850.049505

#### Análisis de Resultados y Rendimiento

Después de escoger algunos de los hiper-parámetros del algoritmo de `XGBoost` por validación cruzada, se usa el modelo para mirar como predice tanto en el entrenamiento como en la prueba y los resultados de manera general son:

* Métricas del modelo:
    |       | RMSE | MAE   |
    | :---        |    ----:   |          ---: |
    | ENTRENAMIENTO      |    79.35    |   54.74     |
    | PRUEBA   |    104.87      |   72.84    |
    | DIFERENCIA PORCENTUAL   |    32.17%      |   33.07%    |
    
    Se puede que realmente hay un overfitting que a pesar de no ser tan grande sí es un poco considerable. El error promedio (mírese el RMSE and MAE) es realmente alto, por ejemplo, en promedio se tiene un MAE de 72.84 unidades vendidas! Una vez ví estos resultados y dándole un poco de sentido al problema, me dí cuenta de lo difícil que es resolver este problema por la variedad de ítems y de variables de estos que se presentan en el marketplace.

    Una especie de __round__ de las predicciones: Sabiendo que `ITEM_SOLD_QUANTITY` realmente es una variable de rangos como se explica [aquí](https://developers.mercadolibre.com.ar/en_us/items-and-searches) se propone una hacer una aproximación a cada predicción al valor más cercano de los posibles que se registraron en la muestra tomada. Después de hacer esto, los resultados mejoraron un poco: 
    
    |       | RMSE | MAE   |
    | :---        |    ----:   |          ---: |
    | ENTRENAMIENTO      |    86.30    |   49.49     |
    | PRUEBA   |    111.63      |   69.39    |
    | DIFERENCIA PORCENTUAL   |    29.35%      |   40.21%    |
    * con aproximación
    
    De manera general el MAE bajo (aunque con la diferencia porcentual el overfitting subió un poco) y el RMSE subió (aunque con la diferencia porcentual el overfitting bajó un poco).
    

* Si miramos los errores para cada cantidad vendida en los valores reales se puede ver que para las cantidades más pequeñas hay un error menor, mientras que para las más grandes el error sube considerablemente.

### NNs (Redes Neuronales)

Uno de los algoritmos que escogí para usar es Redes Neuronales utilizando `Tensorflow` y en particular el modelo de `Keras` dentro de este. 

In [29]:
# Semilla aleatoria fijada 
tf.random.set_seed(200)

# Capa General 1: Capa Convolucional
nn_model = tf.keras.models.Sequential()

# Capa de 10 neuronas con activacion relu y dropout de 0.25
nn_model.add(tf.keras.layers.Dense(10, input_shape=(X_train_total.shape[1],), activation='relu'))
nn_model.add(tf.keras.layers.Dropout(0.25))

# Capa de 25 neuronas con activacion relu y dropout de 0.25
nn_model.add(tf.keras.layers.Dense(25, activation='relu'))
nn_model.add(tf.keras.layers.Dropout(0.25))

# Capa de 25 neuronas con activacion relu y dropout de 0.5
nn_model.add(tf.keras.layers.Dense(25, activation='relu'))
nn_model.add(tf.keras.layers.Dropout(0.5))

# Capa de 25 neuronas con activacion relu y dropout de 0.25
nn_model.add(tf.keras.layers.Dense(25, activation='relu'))
nn_model.add(tf.keras.layers.Dropout(0.25))

# Capa de 10 neuronas con activacion relu y dropout de 0.25
nn_model.add(tf.keras.layers.Dense(10, activation='relu'))
nn_model.add(tf.keras.layers.Dropout(0.25))

# Capa de output 
nn_model.add(tf.keras.layers.Dense(1))

# Se especifica cual es el optimizador (metodo de descenso del gradiente a usar, la función de perdida y las metricas a mostrar)
nn_model.compile(optimizer='adam',loss=tf.keras.losses.MeanSquaredError(), metrics=['mean_absolute_error'])

In [30]:
# Printea la arquitectura de la red
nn_model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_12 (Dense)             (None, 10)                11650     
_________________________________________________________________
dropout_10 (Dropout)         (None, 10)                0         
_________________________________________________________________
dense_13 (Dense)             (None, 25)                275       
_________________________________________________________________
dropout_11 (Dropout)         (None, 25)                0         
_________________________________________________________________
dense_14 (Dense)             (None, 25)                650       
_________________________________________________________________
dropout_12 (Dropout)         (None, 25)                0         
_________________________________________________________________
dense_15 (Dense)             (None, 25)               

In [31]:
# Se hace el fit (detras un metodo de gradiente)
nn_model.fit(X_train_total, Y_train, epochs=250, batch_size=2000, verbose=0, validation_split=0.1)

In [32]:
# Se generan las predicciones para el entrenamiento y prueba y se evaluan dos metricas: RMSE y MAE
test_predictions = nn_model.predict(X_test_total)
train_predictions = nn_model.predict(X_train_total)

In [33]:
# Comparacion RSME train y test
rmse_train = np.sqrt(mean_squared_error(y_pred=train_predictions, y_true=Y_train))
rmse_test = np.sqrt(mean_squared_error(y_pred=test_predictions, y_true=Y_test))
print('RMSE DE ENTRENAMIENTO: {}'.format(np.round(rmse_train,2)))
print('RMSE DE PRUEBA: {}'.format(np.round(rmse_test,2)))
diferencia_porcentual = np.round(100*(rmse_test-rmse_train)/rmse_train,2)
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

RMSE DE ENTRENAMIENTO: 101.38
RMSE DE PRUEBA: 112.71
DIFERENCIA DE: 11.17%


In [34]:
# Comparacion MAE train y test
mae_train = mean_absolute_error(y_pred=train_predictions, y_true=Y_train)
mae_test = mean_absolute_error(y_pred=test_predictions, y_true=Y_test)
diferencia_porcentual = np.round(100*(mae_test-mae_train)/mae_train,2)
print('MAE DE ENTRENAMIENTO: {}'.format(np.round(mae_train,2)))
print('MAE DE PRUEBA: {}'.format(np.round(mae_test,2)))
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

MAE DE ENTRENAMIENTO: 67.57
MAE DE PRUEBA: 75.05
DIFERENCIA DE: 11.07%


In [35]:
values = sorted(Y_train.unique())
test_predictions = [find_nearest(values, p)for p in test_predictions] 
train_predictions = [find_nearest(values, p)for p in train_predictions] 

In [36]:
# Comparacion RSME train y test
rmse_train = np.sqrt(mean_squared_error(y_pred=train_predictions, y_true=Y_train))
rmse_test = np.sqrt(mean_squared_error(y_pred=test_predictions, y_true=Y_test))
print('RMSE DE ENTRENAMIENTO: {}'.format(np.round(rmse_train,2)))
print('RMSE DE PRUEBA: {}'.format(np.round(rmse_test,2)))
diferencia_porcentual = np.round(100*(rmse_test-rmse_train)/rmse_train,2)
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

RMSE DE ENTRENAMIENTO: 110.47
RMSE DE PRUEBA: 118.93
DIFERENCIA DE: 7.66%


In [37]:
# Comparacion MAE train y test
mae_train = mean_absolute_error(y_pred=train_predictions, y_true=Y_train)
mae_test = mean_absolute_error(y_pred=test_predictions, y_true=Y_test)
diferencia_porcentual = np.round(100*(mae_test-mae_train)/mae_train,2)
print('MAE DE ENTRENAMIENTO: {}'.format(np.round(mae_train,2)))
print('MAE DE PRUEBA: {}'.format(np.round(mae_test,2)))
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

MAE DE ENTRENAMIENTO: 68.29
MAE DE PRUEBA: 74.07
DIFERENCIA DE: 8.47%


In [38]:
# Se crea un dataframe con los resultados
resultados_train = pd.DataFrame({'TYPE': ['TRAIN']*len(Y_train), 'TRUE': Y_train, 'PREDICTED': train_predictions})
resultados_test = pd.DataFrame({'TYPE': ['TEST']*len(Y_test), 'TRUE': Y_test, 'PREDICTED': test_predictions})
resultados_nn = pd.concat([resultados_train, resultados_test], axis=0).reset_index(drop=True)
resultados_nn['ABSOLUTE_ERROR'] = (resultados_nn['TRUE']-resultados_nn['PREDICTED']).abs()
resultados_nn['SQUARE_ERROR'] = (resultados_nn['TRUE']-resultados_nn['PREDICTED'])**2

In [39]:
# Resultados por cantidad
resultados_por_cantidad = resultados_nn.groupby(['TYPE', 'TRUE'])[['ABSOLUTE_ERROR', 'SQUARE_ERROR']].mean().reset_index()

In [40]:
# Imprime los resultados por cantidad en cuanto al MAE y MSE para los datos de prueba
resultados_por_cantidad.loc[resultados_por_cantidad['TYPE']=='TEST']

TYPE  TRUE  ABSOLUTE_ERROR  SQUARE_ERROR
0   TEST     0       32.572327   1091.150943
1   TEST     1       31.567164   1077.268657
2   TEST     2       30.506494   1005.077922
3   TEST     3       32.562500   1650.562500
4   TEST     4       27.738462    782.169231
5   TEST     5       37.769106   2408.884553
6   TEST     6       94.000000   8836.000000
7   TEST     7       93.000000   8649.000000
8   TEST     8       27.000000    729.000000
9   TEST    14       45.000000   2025.000000
10  TEST    25       35.348214   2749.839286
11  TEST    50       38.716094   3252.368897
12  TEST    56       94.000000   8836.000000
13  TEST   100       50.536023   4336.599424
14  TEST   150       51.550218   4097.890830
15  TEST   200       71.705128   8198.346154
16  TEST   250       70.143389  10361.778399
17  TEST   500      227.262376  67745.268977

#### Análisis de Resultados y Rendimiento

Después de correr un `MLP` con una arquitectura muy simple (algunas capas intermedias con activación `relu` y `dropout`) se usa el modelo para mirar como predice tanto en el entrenamiento como en la prueba y los resultados de manera general son:


* Métricas del modelo:
    |       | RMSE | MAE   |
    | :---        |    ----:   |          ---: |
    | ENTRENAMIENTO      |    101.38    |   67.57     |
    | PRUEBA   |    112.71      |    75.05   |
    | DIFERENCIA PORCENTUAL   |    11.17%     |    11.07%    |
    
    Se puede que realmente hay un overfitting mucho menor en porcentualidad al encontrado en el algoritmo `XGBoost`. El error promedio (mírese el RMSE and MAE) es realmente alto, por ejemplo, en promedio se tiene un MAE de 75.05 unidades vendidas!

    De nuevo, después de hacer la aproximación, los resultados mejoraron un poco: 
    
    |       | RMSE | MAE   |
    | :---        |    ----:   |          ---: |
    | ENTRENAMIENTO      |    110.47    |   68.29     |
    | PRUEBA   |    118.93      |   74.07    |
    | DIFERENCIA PORCENTUAL   |    7.66%      |   8.47%   |
    * con aproximación
    
    De manera general el MAE bajo y el RMSE subió. Para ambos tipos de errores con la diferencia porcentual el overfitting bajó un poco.
    

* Si miramos los errores para cada cantidad vendida en los valores reales se puede ver que los errores promedio son más similares con respecto a lo obtenido con el `XGBoost`.

### Random Forest (Bosques Aleatorios)

Uno de los algoritmos que escogí para usar es Bosques Aleatorios utilizando `ScikitLearn`. 

In [158]:
# Diccionario de hiperparametros a tunear
param_tuning = {'n_estimators': [100],
               'max_features': ['sqrt', 'log2'],
               'max_depth': [7, 10],
               'min_samples_leaf': [2, 4],
               'min_samples_split': [5, 10],
               'bootstrap': [False]}

# Se evalua para todas las combinaciones de hiperparametros cual tiene el menor error de validacion
# usando validacion cruzada para 5-folds

# Modelo RFRegressor
rf_model = RandomForestRegressor(random_state=1608)

# Validacion cruzada
k_folds = KFold(n_splits=5, shuffle=True, random_state=1608)
grid_search_cv = GridSearchCV(estimator = rf_model,
                              param_grid = param_tuning,            
                              cv = k_folds,
                              n_jobs = -1,
                              verbose = 1)
cv_fit = grid_search_cv.fit(X=X_train_total, y=Y_train) # Fit

Fitting 5 folds for each of 16 candidates, totalling 80 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 16 concurrent workers.
[Parallel(n_jobs=-1)]: Done  18 tasks      | elapsed:   15.4s
[Parallel(n_jobs=-1)]: Done  80 out of  80 | elapsed:   42.9s finished


In [159]:
# Imprime los mejores parametros bajo el score de cv calculado para cada iteracion
grid_search_cv.best_params_

{'bootstrap': False,
 'max_depth': 10,
 'max_features': 'sqrt',
 'min_samples_leaf': 2,
 'min_samples_split': 5,
 'n_estimators': 100}

In [160]:
# Se generan las predicciones para el entrenamiento y prueba y se evaluan dos metricas: RMSE y MAE
rf_best_model = grid_search_cv.best_estimator_
test_predictions = rf_best_model.predict(X_test_total)
train_predictions = rf_best_model.predict(X_train_total)

In [161]:
# Comparacion RSME train y test
rmse_train = np.sqrt(mean_squared_error(y_pred=train_predictions, y_true=Y_train))
rmse_test = np.sqrt(mean_squared_error(y_pred=test_predictions, y_true=Y_test))
print('RMSE DE ENTRENAMIENTO: {}'.format(np.round(rmse_train,2)))
print('RMSE DE PRUEBA: {}'.format(np.round(rmse_test,2)))
diferencia_porcentual = np.round(100*(rmse_test-rmse_train)/rmse_train,2)
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

RMSE DE ENTRENAMIENTO: 137.27
RMSE DE PRUEBA: 139.83
DIFERENCIA DE: 1.86%


In [162]:
# Comparacion MAE train y test
mae_train = mean_absolute_error(y_pred=train_predictions, y_true=Y_train)
mae_test = mean_absolute_error(y_pred=test_predictions, y_true=Y_test)
diferencia_porcentual = np.round(100*(mae_test-mae_train)/mae_train,2)
print('MAE DE ENTRENAMIENTO: {}'.format(np.round(mae_train,2)))
print('MAE DE PRUEBA: {}'.format(np.round(mae_test,2)))
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

MAE DE ENTRENAMIENTO: 108.64
MAE DE PRUEBA: 111.3
DIFERENCIA DE: 2.44%


In [50]:
values = sorted(Y_train.unique())
test_predictions = [find_nearest(values, p)for p in test_predictions] 
train_predictions = [find_nearest(values, p)for p in train_predictions] 

In [51]:
# Comparacion RSME train y test
rmse_train = np.sqrt(mean_squared_error(y_pred=train_predictions, y_true=Y_train))
rmse_test = np.sqrt(mean_squared_error(y_pred=test_predictions, y_true=Y_test))
print('RMSE DE ENTRENAMIENTO: {}'.format(np.round(rmse_train,2)))
print('RMSE DE PRUEBA: {}'.format(np.round(rmse_test,2)))
diferencia_porcentual = np.round(100*(rmse_test-rmse_train)/rmse_train,2)
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

RMSE DE ENTRENAMIENTO: 140.18
RMSE DE PRUEBA: 142.21
DIFERENCIA DE: 1.45%


In [52]:
# Comparacion MAE train y test
mae_train = mean_absolute_error(y_pred=train_predictions, y_true=Y_train)
mae_test = mean_absolute_error(y_pred=test_predictions, y_true=Y_test)
diferencia_porcentual = np.round(100*(mae_test-mae_train)/mae_train,2)
print('MAE DE ENTRENAMIENTO: {}'.format(np.round(mae_train,2)))
print('MAE DE PRUEBA: {}'.format(np.round(mae_test,2)))
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

MAE DE ENTRENAMIENTO: 109.64
MAE DE PRUEBA: 111.57
DIFERENCIA DE: 1.76%


In [53]:
# Se crea un dataframe con los resultados
resultados_train = pd.DataFrame({'TYPE': ['TRAIN']*len(Y_train), 'TRUE': Y_train, 'PREDICTED': train_predictions})
resultados_test = pd.DataFrame({'TYPE': ['TEST']*len(Y_test), 'TRUE': Y_test, 'PREDICTED': test_predictions})
resultados_rf = pd.concat([resultados_train, resultados_test], axis=0).reset_index(drop=True)
resultados_rf['ABSOLUTE_ERROR'] = (resultados_rf['TRUE']-resultados_rf['PREDICTED']).abs()
resultados_rf['SQUARE_ERROR'] = (resultados_rf['TRUE']-resultados_rf['PREDICTED'])**2

In [54]:
# Resultados por cantidad
resultados_por_cantidad = resultados_rf.groupby(['TYPE', 'TRUE'])[['ABSOLUTE_ERROR', 'SQUARE_ERROR']].mean().reset_index()

In [55]:
# Imprime los resultados por cantidad en cuanto al MAE y MSE para los datos de prueba
resultados_por_cantidad.loc[resultados_por_cantidad['TYPE']=='TEST']

TYPE  TRUE  ABSOLUTE_ERROR  SQUARE_ERROR
0   TEST     0       65.811321   5248.113208
1   TEST     1       75.238806   6200.194030
2   TEST     2       82.038961   7453.545455
3   TEST     3       81.541667   7255.958333
4   TEST     4       81.938462   7408.707692
5   TEST     5      105.443902  12439.856911
6   TEST     6      119.000000  14786.000000
7   TEST     7       93.000000   8649.000000
8   TEST     8      142.000000  20164.000000
9   TEST    14       86.000000   7396.000000
10  TEST    25      103.375000  11818.209821
11  TEST    50       92.593128   9869.775769
12  TEST    56       94.000000   8836.000000
13  TEST   100       55.778098   4570.331412
14  TEST   150       26.816594   1662.799127
15  TEST   200       39.102564   2564.102564
16  TEST   250       63.314711   5623.836127
17  TEST   500      291.899340  86834.952145

#### Análisis de Resultados y Rendimiento

Después de escoger por la validación cruzada los hiper-parámetros del `RandomForestRegressor` se usa el modelo para mirar como predice tanto en el entrenamiento como en la prueba y los resultados de manera general son:

* Métricas del modelo:
    |       | RMSE | MAE   |
    | :---        |    ----:   |          ---: |
    | ENTRENAMIENTO      |    137.27    |   108.64     |
    | PRUEBA   |    139.83      |    111.3   |
    | DIFERENCIA PORCENTUAL   |    1.86%     |    2.44%    |
    
    Se puede que realmente hay un overfitting mucho menor en porcentualidad al encontrado en los algoritmos anterior. Sin embargo, el error promedio (mírese el RMSE and MAE) es demasiado altos: en promedio se tiene un MAE de 108.64 unidades vendidas, más de 100 unidades vendidas!

    De nuevo, después de hacer la aproximación, los resultados mejoraron un poco: 
    
    |       | RMSE | MAE   |
    | :---        |    ----:   |          ---: |
    | ENTRENAMIENTO      |    140.18    |   109.64     |
    | PRUEBA   |    142.21      |   111.57    |
    | DIFERENCIA PORCENTUAL   |    1.45%      |   1.76%  |
    * con aproximación

    De manera general el MAE bajo y el RMSE subió. Para ambos tipos de errores con la diferencia porcentual el overfitting bajó un poco.
    

* Si miramos los errores para cada cantidad vendida en los valores reales se puede ver que los errores promedio son mucho peores que los registrados en los anteriores modelos.

Este fue el peor de los algoritmos evaluados por las métricas de error, además de ser el que más tiempo tomó (y eso que la validación cruzada se hizo para muy pocas combinaciones de parámetros). Eso sí, creo que faltó poner muchos parámetros como `max_depth` con mayor complejidad, es decir permitir árboles más extensos en general.

### Stacked Model

Una idea final que tuve para lograr el mejor modelo posible dentro del tiempo que tenía para experimentar con los datos fue hacer un modelo que a partir de las predicciones de los tres modelos presentados anteriormente.

In [56]:
# Construccion de dataset de predicciones
resultados = resultados_xgb[['TYPE', 'TRUE', 'PREDICTED']].rename(columns={'PREDICTED': 'PREDICTED_XGB'})
resultados['PREDICTED_NN'] = resultados_nn['PREDICTED']
resultados['PREDICTED_RF'] = resultados_rf['PREDICTED']

# Separa en train y test
X_train_stack = resultados.loc[resultados['TYPE']=='TRAIN'].drop(columns=['TYPE', 'TRUE'])
X_test_stack = resultados.loc[resultados['TYPE']=='TEST'].drop(columns=['TYPE', 'TRUE'])
Y_train_stack = resultados.loc[resultados['TYPE']=='TRAIN', 'TRUE'] 
Y_test_stack = resultados.loc[resultados['TYPE']=='TEST', 'TRUE']

In [57]:
# Diccionario de hiperparametros a tunear
param_tuning = {'n_estimators': [100, 1000],
               'max_features': ['auto', 'sqrt'],
               'max_depth': [3, 5, 7, 10],
               'bootstrap': [False]}

# Se evalua para todas las combinaciones de hiperparametros cual tiene el menor error de validacion
# usando validacion cruzada para 5-folds

# Modelo RFRegressor
rf_stack_model = RandomForestRegressor(random_state=1608)

# Validacion cruzada
k_folds = KFold(n_splits=5, shuffle=True, random_state=1608)
grid_search_cv = GridSearchCV(estimator = rf_model,
                              param_grid = param_tuning,            
                              cv = k_folds,
                              n_jobs = -1,
                              verbose = 1)
cv_fit = grid_search_cv.fit(X=X_train_stack, y=Y_train_stack) # Fit

Fitting 5 folds for each of 16 candidates, totalling 80 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 16 concurrent workers.
[Parallel(n_jobs=-1)]: Done  18 tasks      | elapsed:    2.8s
[Parallel(n_jobs=-1)]: Done  80 out of  80 | elapsed:   10.8s finished


In [58]:
# Imprime los mejores parametros bajo el score de cv calculado para cada iteracion
grid_search_cv.best_params_

{'bootstrap': False,
 'max_depth': 7,
 'max_features': 'sqrt',
 'n_estimators': 100}

In [59]:
# Se generan las predicciones para el entrenamiento y prueba y se evaluan dos metricas: RMSE y MAE
rf_stack_best_model = grid_search_cv.best_estimator_
test_predictions = rf_stack_best_model.predict(X_test_stack)
train_predictions = rf_stack_best_model.predict(X_train_stack)

In [60]:
# Comparacion RSME train y test
rmse_train = np.sqrt(mean_squared_error(y_pred=train_predictions, y_true=Y_train))
rmse_test = np.sqrt(mean_squared_error(y_pred=test_predictions, y_true=Y_test))
print('RMSE DE ENTRENAMIENTO: {}'.format(np.round(rmse_train,2)))
print('RMSE DE PRUEBA: {}'.format(np.round(rmse_test,2)))
diferencia_porcentual = np.round(100*(rmse_test-rmse_train)/rmse_train,2)
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

RMSE DE ENTRENAMIENTO: 75.7
RMSE DE PRUEBA: 104.95
DIFERENCIA DE: 38.65%


In [61]:
# Comparacion MAE train y test
mae_train = mean_absolute_error(y_pred=train_predictions, y_true=Y_train)
mae_test = mean_absolute_error(y_pred=test_predictions, y_true=Y_test)
diferencia_porcentual = np.round(100*(mae_test-mae_train)/mae_train,2)
print('MAE DE ENTRENAMIENTO: {}'.format(np.round(mae_train,2)))
print('MAE DE PRUEBA: {}'.format(np.round(mae_test,2)))
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

MAE DE ENTRENAMIENTO: 47.91
MAE DE PRUEBA: 67.06
DIFERENCIA DE: 39.98%


In [62]:
values = sorted(Y_train.unique())
test_predictions = [find_nearest(values, p)for p in test_predictions] 
train_predictions = [find_nearest(values, p)for p in train_predictions] 

In [63]:
# Comparacion RSME train y test
rmse_train = np.sqrt(mean_squared_error(y_pred=train_predictions, y_true=Y_train))
rmse_test = np.sqrt(mean_squared_error(y_pred=test_predictions, y_true=Y_test))
print('RMSE DE ENTRENAMIENTO: {}'.format(np.round(rmse_train,2)))
print('RMSE DE PRUEBA: {}'.format(np.round(rmse_test,2)))
diferencia_porcentual = np.round(100*(rmse_test-rmse_train)/rmse_train,2)
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

RMSE DE ENTRENAMIENTO: 82.66
RMSE DE PRUEBA: 111.73
DIFERENCIA DE: 35.17%


In [64]:
# Comparacion MAE train y test
mae_train = mean_absolute_error(y_pred=train_predictions, y_true=Y_train)
mae_test = mean_absolute_error(y_pred=test_predictions, y_true=Y_test)
diferencia_porcentual = np.round(100*(mae_test-mae_train)/mae_train,2)
print('MAE DE ENTRENAMIENTO: {}'.format(np.round(mae_train,2)))
print('MAE DE PRUEBA: {}'.format(np.round(mae_test,2)))
print('DIFERENCIA DE: {}%'.format(diferencia_porcentual))

MAE DE ENTRENAMIENTO: 43.99
MAE DE PRUEBA: 64.83
DIFERENCIA DE: 47.38%


In [65]:
# Se crea un dataframe con los resultados
resultados_train = pd.DataFrame({'TYPE': ['TRAIN']*len(Y_train), 'TRUE': Y_train, 'PREDICTED': train_predictions})
resultados_test = pd.DataFrame({'TYPE': ['TEST']*len(Y_test), 'TRUE': Y_test, 'PREDICTED': test_predictions})
resultados_rf = pd.concat([resultados_train, resultados_test], axis=0).reset_index(drop=True)
resultados_rf['ABSOLUTE_ERROR'] = (resultados_rf['TRUE']-resultados_rf['PREDICTED']).abs()
resultados_rf['SQUARE_ERROR'] = (resultados_rf['TRUE']-resultados_rf['PREDICTED'])**2

In [66]:
# Resultados por cantidad
resultados_por_cantidad = resultados_rf.groupby(['TYPE', 'TRUE'])[['ABSOLUTE_ERROR', 'SQUARE_ERROR']].mean().reset_index()

In [67]:
# Imprime los resultados por cantidad en cuanto al MAE y MSE para los datos de prueba
resultados_por_cantidad.loc[resultados_por_cantidad['TYPE']=='TEST']

TYPE  TRUE  ABSOLUTE_ERROR  SQUARE_ERROR
0   TEST     0        8.710692    176.283019
1   TEST     1       10.164179    313.208955
2   TEST     2        7.909091    129.285714
3   TEST     3       11.354167    442.020833
4   TEST     4        7.646154    159.492308
5   TEST     5       26.447154   2231.744715
6   TEST     6       69.000000   5386.000000
7   TEST     7       48.000000   2304.000000
8   TEST     8       15.000000    225.000000
9   TEST    14       17.000000    289.000000
10  TEST    25       38.232143   3707.263393
11  TEST    50       51.737794   6026.287523
12  TEST    56       44.000000   1936.000000
13  TEST   100       71.368876  10630.792507
14  TEST   150       72.834061  11148.414847
15  TEST   200       79.346154  10707.512821
16  TEST   250       86.249534  16807.702048
17  TEST   500      144.503300  41973.665017

#### Análisis de Resultados y Rendimiento

Se hizo un _stacking_ o apilamiento de las predicciones de los modelos usados y se usó como modelo un simple `RandomForestRegressor` y obtuvieron los siguientes resultados:

* Métricas del modelo:
    |       | RMSE | MAE   |
    | :---        |    ----:   |          ---: |
    | ENTRENAMIENTO      |    75.7    |   47.91     |
    | PRUEBA   |    104.95      |    67.06   |
    | DIFERENCIA PORCENTUAL   |    38.65%     |    39.98%    |
    
    A pesar de haber un overfitting notable, es claro que, sin contar los resultados de aproximación, el __Stack Model__ es el que mejores métricas de error tiene.

    De nuevo, después de hacer la aproximación, los resultados mejoraron un poco: 
    
    |       | RMSE | MAE   |
    | :---        |    ----:   |          ---: |
    | ENTRENAMIENTO      |    82.66    |   43.99     |
    | PRUEBA   |    111.73      |   64.83    |
    | DIFERENCIA PORCENTUAL   |    35.17%      |   47.38%  |
    * con aproximación

    A pesar de tener un overfitting fuerte (no corregido por la aproximación) si es el modelo o la predicción con las mejores métricas de error.
    
* Si miramos los errores para cada cantidad vendida en los valores reales se puede ver que sigue los mismos comportamientos que el `XGBoost`.

Este fue el mejor de los algoritmos evaluados por las métricas de error.

## Conclusiones y Reflexiones

Lo primero que se debe resaltar es la dificultad tan fuerte en generar un modelo que estime cuando se tiene una variedad tan diferente de ítems, vendedores y demás, sin mencionar que el conjunto de datos no es más que una muestra (y no comprende el total de ítems de [Mercado Libre Colombia](https://www.mercadolibre.com.co/)). 

Otro aspecto que creo que puede ser muy relevante es que tal vez, por la construcción del dataset, se tienen muchos ítems pérdidos en el MarketPlace (con un parámetro `offset` muy alto) y tal vez estos esten haciendo algún tipo de ruido en el modelo.

El modelo construido usando el concepto de stack/apilar efectivamente resulta ser el mejor cuando se examinan las métricas (de prueba) de RMSE y MAE, lo cual para mí tiene mucho sentido pues junta y combina la __información__ (las predicciones) de cada uno de los modelos individuales. 

La cantidad vendida realmente corresponde a un set de rangos establecidos [aquí](https://developers.mercadolibre.com.ar/en_us/items-and-searches) y además algunos ítems tienen valores que no se presentan en el cuadro que muestra los valores de `sold_quantity` y por lo tanto, hay que añadir que por eso hay un poco de ruído en la variable de respuesta. El problema es demasiado retador y sería interesante ver que resultasdos se obtiene con mucho más tiempo y conocimiento del marketplace y de las features.

Veo un potencial en construir algunas variables (feature engineering) con un poco más de tiempo e imaginación, además de usar tanto el texto del ítem como la imagen de este que efectivamente se puede obtener con la información del conjunto de datos (ver el código anexo). Sin embargo, el uso de imágenes y texto requeriría una muestra más grande de productos (tal vez de otros países también).

_Una posibilidad_...

__Modelo Alternativo.__ Se propone es trabajar con rangos de `ITEM_SOLD_QUANTITY`. Por ejemplo, se podrían usar los siguientes rangos (que serán tomados como clases):

1. 0-50: De 0 a 50 cantdiades vendidas.
2. 51-100: De 51 a 100 cantdiades vendidas.
3. 101-150: De 101 a 150 cantdiades vendidas.
4. 151-200: De 151 a 200 cantdiades vendidas.
5. 201-250: De 201 a 250 cantdiades vendidas.
6. 251-500: De 251 a 500 cantdiades vendidas.

In [880]:
# Descarga la imagen con la url de un item
img_url_example = products.loc[1608, 'ITEM_THUMBNAIL']
response = requests.get(img_url_example)

# Guarda la imagen
print('DESCARGADA IMAGEN DE: {}'.format(products.loc[1608, 'ITEM_TITLE'].upper()))
file = open("../Data/Images/EXAMPLE.png", "wb")
file.write(response.content)
file.close()

DESCARGADA IMAGEN DE: LLAVERO LUJO TRENZA CUERO MARCA TOYOTA PRADO TXL FORTU CRUIS


<img src="../Data/Images/EXAMPLE.png " alt="drawing" width="200" title='Ejemplo de la miniatura de un ítem'/>
<center>  <font size="1"> Ejemplo de la miniatura de un ítem del market place de Mercado Libre Colombia</font> </center>

# Fin del Documento